{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# bQTL distance to motif vs p-value\n",
    "- Take the bQTL SNPs from supplemental material section of bQTL paper (mmc2.xlsx). There are four tabs in the xlsx, one for each TF (JUND/RELA/SPI1/STAT1). Each entry in the table contains the location and p-value for an SNP, among other info.\n",
    "- For each TF, take first 100k bQTLs of smallest pvals (there are too many SNPs)\n",
    "- Calculate the distance between each bQTL SNP to its nearest CIS-BP motif center. As the CIS-BP motifs do not come with weights, the motif center is simply (start+end)/2\n",
    "- Cap the distance at 1000bp and 10,000bp and plot the distance vs -log10(p-value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function, division\n",
    "\n",
    "import sys\n",
    "import os\n",
    "#import matplotlib as mpl\n",
    "#mpl.use('Agg')\n",
    "\n",
    "import numpy as np\n",
    "import sys\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "MAX_LOC = 250000000L\n",
    "def chrom_to_idx(chrom, loc):\n",
    "    chrid = chrom[3:]\n",
    "    if chrid == 'x' or chrid == 'X':\n",
    "        chrnm = 23\n",
    "    elif chrid == 'y' or chrid == 'Y':\n",
    "        chrnm = 24\n",
    "    else:\n",
    "        chrnm = int(chrid)\n",
    "    return long(chrnm * MAX_LOC + loc)\n",
    "\n",
    "def idx_to_chrom(idx):\n",
    "    loc = idx % MAX_LOC\n",
    "    chrid = int(idx / MAX_LOC)\n",
    "    if chrid == 23:\n",
    "        chrom = 'chrX'\n",
    "    elif chrid == 24:\n",
    "        chrom = 'chrY'\n",
    "    else:\n",
    "        chrom = 'chr' + str(chrid)\n",
    "    return chrom, loc\n",
    "        \n",
    "    \n",
    "def calc_seq_loc(seq):\n",
    "    st = seq.coor.start\n",
    "    en = seq.coor.end\n",
    "    local = int((st + en)/2) # fix later using center of gravity\n",
    "\n",
    "    tsv = merged_tsv_list[seq.coor.example_idx]\n",
    "    location = int(tsv[1]) + local\n",
    "    if location > MAX_LOC:\n",
    "        print(\"location=%d, tsv_start=%d, ex=%d, st=%d, en=%d\" %(location, int(tsv[1]), seq.coor.example_idx, seq.coor.start, seq.coor.end))\n",
    "        print(tsv)\n",
    "    idx = chrom_to_idx(tsv[0], location)\n",
    "\n",
    "    return idx, tsv[0], location\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JUND done\n",
      "RELA done\n",
      "SPI1 done\n",
      "STAT1 done\n"
     ]
    }
   ],
   "source": [
    "from collections import OrderedDict\n",
    "tf_list = [ 'JUND', 'RELA', 'SPI1', 'STAT1']\n",
    "\n",
    "tf_dict = OrderedDict()\n",
    "\n",
    "for tf in tf_list:\n",
    "    #with open(\"data/\" + tf + \"_sort.txt\") as in_fh:\n",
    "    with open(\"under_1ksummit/\" + tf + \"_sort.txt\") as in_fh:\n",
    "\n",
    "        lines = [ map(int, line.split()) for line in in_fh ]\n",
    "        tf_dict[tf] = list(zip(*lines)) # transpose\n",
    "        print(tf + \" done\")\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from bisect import bisect_left\n",
    "max_chrid = 25\n",
    "bounds  = [0] * (max_chrid+1)\n",
    "\n",
    "\n",
    "def find_nearest(my_list, my_idx):\n",
    "    \"\"\"\n",
    "    Assumes my_list is sorted. Returns closest value to my_idx.\n",
    "    If two numbers are equally close, return the smallest number.\n",
    "    If lo\n",
    "    \"\"\"\n",
    "\n",
    "    chrid = long(my_idx / MAX_LOC)\n",
    "    lo = bounds[chrid-1]\n",
    "    hi = bounds[chrid]\n",
    "    if lo == hi:\n",
    "        return -1, -1 # this chrom is empty\n",
    "    \n",
    "    # lo != hi, so the chrom is not empty\n",
    "    pos = bisect_left(my_list, my_idx, lo=lo, hi=hi)\n",
    "\n",
    "    if pos == lo:\n",
    "        return lo, my_list[lo]\n",
    "    if pos == hi:\n",
    "        return hi-1, my_list[hi-1] # hi belongs to next chrom\n",
    "    before = my_list[pos - 1]\n",
    "    after = my_list[pos]\n",
    "    #print(lo, hi, pos, before, after)\n",
    "    if after - my_idx < my_idx - before:\n",
    "       return pos, after\n",
    "    else:\n",
    "       return pos-1, before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "\n",
    "for tf in tf_list:\n",
    "\n",
    "    idx_array = tf_dict[tf][1]\n",
    "    for chrid in range(1, max_chrid+1):\n",
    "        lo_val = chrid * MAX_LOC\n",
    "        bounds[chrid-1] = bisect_left(idx_array, lo_val)\n",
    "        #print(chrid-1, lo_val, bounds[chrid-1])\n",
    "\n",
    "    #print(idx_array[:10])\n",
    "    #print(bounds)\n",
    "\n",
    "\n",
    "    #snp_dir = \"/Users/kat/kundajelab/tmp/bQTL/bQTL_all_SNPs/\"\n",
    "    snp_dir = \"/home/ktian/kundajelab/tfnet/results/nandi/bQTL/analysis/bQTL_all_SNPs/\"\n",
    "    snp_list = []\n",
    "    snp_file = tf + '.txt'\n",
    "    #snp_file = tf + '.txt'\n",
    "\n",
    "    with open(snp_dir + snp_file) as in_fh:\n",
    "        header = next(in_fh)\n",
    "        line_num = 0\n",
    "        for line in in_fh:\n",
    "            fields = line.split('\\t')\n",
    "            snp_chrom = fields[0]\n",
    "            snp_pos   = int(fields[1])\n",
    "            pval  = float(fields[9])\n",
    "            idx   = chrom_to_idx(snp_chrom, snp_pos)\n",
    "            motif_off, motif_idx = find_nearest(idx_array, idx)\n",
    "            if motif_off >= 0:\n",
    "                snp_list.append([abs(idx-motif_idx), -math.log(pval,10)])\n",
    "                #snp_list.append([abs(idx-motif_idx), pval])\n",
    "            line_num += 1\n",
    "    #print(snp_list[:5])     \n",
    "    #print(len(snp_list))\n",
    "\n",
    "    with open(\"under_1ksummit/\" + tf + \"_cisbp_full.tsv\", \"w\") as fh:\n",
    "        for row in snp_list:\n",
    "            fh.write(str(row[0]) +\"\\t\"+ str(row[1]) + \"\\n\")\n",
    "\n",
    "    \"\"\"\n",
    "    %matplotlib inline\n",
    "\n",
    "    import matplotlib.pyplot as plt\n",
    "    import numpy as np\n",
    "    import matplotlib\n",
    "    from scipy.stats import gaussian_kde\n",
    "\n",
    "    #fig=plt.figure(figsize=(10, 8), dpi= 100)\n",
    "    snp_dist = [item[0] for item in snp_list if item[1] > 0 and item[0] <= 1000]\n",
    "    snp_pval = [item[1] for item in snp_list if item[1] > 0 and item[0] <= 1000]\n",
    "\n",
    "    x=snp_dist\n",
    "    y=snp_pval\n",
    "    xy = np.vstack([x,y])\n",
    "    z = gaussian_kde(xy)(xy)\n",
    "\n",
    "    print(\"Of top 100k \"+tf+\" SNPs with highest p-val, \", len(snp_dist), \" SNPs' distances to the nearest motif center are < 1000\")\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_axes([1,1,1,1])\n",
    "    plt.scatter(snp_dist, snp_pval, 1, c=z, alpha=1, marker='o', label=\".\")\n",
    "    plt.xlabel(\"SNP distance to \"+ tf +\" motif centers, \" + str(len(snp_dist)) + \" SNPs\")\n",
    "    plt.ylabel(\"SNP -log10(pval)\")\n",
    "    plt.colorbar(label='density (low to high)')\n",
    "\n",
    "    #plt.legend(loc=2)\n",
    "    plt.show()    \n",
    "    fig.savefig(\"under_1ksummit/\" + tf + \"_cisbp_1k_full.png\", bbox_inches='tight')\n",
    "\n",
    "    # https://stackoverflow.com/questions/20105364/how-can-i-make-a-scatter-plot-colored-by-density-in-matplotlib\n",
    "    \n",
    "\n",
    "    \n",
    "    %matplotlib inline\n",
    "\n",
    "    import matplotlib.pyplot as plt\n",
    "    import numpy as np\n",
    "    import matplotlib\n",
    "    from scipy.stats import gaussian_kde\n",
    "\n",
    "    for item in snp_list:\n",
    "        if item[0] == 0:\n",
    "            item[0] = 0.1\n",
    "    #fig=plt.figure(figsize=(10, 8), dpi= 100)\n",
    "    snp_dist = [item[0] for item in snp_list]\n",
    "    snp_pval = [item[1] for item in snp_list]\n",
    "\n",
    "    x=snp_dist\n",
    "    y=snp_pval\n",
    "    xy = np.vstack([x,y])\n",
    "    z = gaussian_kde(xy)(xy)\n",
    "\n",
    "    print(\"Of top 100k \"+tf+\" SNPs with highest p-val, \", len(snp_dist), \" SNPs have motif in the same Chromosome\")\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_axes([1,1,1,1])\n",
    "    plt.scatter(snp_dist, snp_pval, 1, c=z, alpha=1, marker='o', label=\".\")\n",
    "    plt.xscale('log')\n",
    "    #plt.yscale('log')\n",
    "    plt.xlabel(\"SNP distance to \"+ tf +\" motif centers, \" + str(len(snp_dist)) + \" SNPs\")\n",
    "    plt.ylabel(\"SNP -log10(pval)\")\n",
    "    plt.colorbar(label='density (low to high)')\n",
    "\n",
    "    #plt.legend(loc=2)\n",
    "    plt.show()    \n",
    "    fig.savefig(\"under_1ksummit/\" + tf + \"_cisbp_log_big.png\", bbox_inches='tight')\n",
    "    \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  4361 under_1ksummit/JUND_sort.txt\r\n",
      "  8917 under_1ksummit/RELA_sort.txt\r\n",
      " 10747 under_1ksummit/SPI1_sort.txt\r\n",
      "  1838 under_1ksummit/STAT1_sort.txt\r\n",
      " 25863 total\r\n"
     ]
    }
   ],
   "source": [
    "!wc -l under_1ksummit/*sort.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "above are the number of motifs for each TF in the CIS-BP database. The nearest motif center for a SNP is only calculated among motifs for the same TF. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "modisco_dev",
   "language": "python",
   "name": "modisco_dev"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
